enzyme/Enzyme/MLIR/Passes/EnzymeBatchDiffPass.cpp

```diff
@@ -75,377 +316,798 @@ namespace {
 
-    OpBuilder builder(op);
+      for (auto &pair : toMerge) {
+        auto key = pair.first;
+        auto allDiffs = pair.second;
+        if (allDiffs.size() < 2)
+          continue;
+
+        // Collect inner effects of function
+        SmallVector<MemoryEffects::EffectInstance> calleeEffects =
+            batchutils::collectFnEffects(innerEffectCache, key.function);
+
+        // TODO skip if known readonly from existing analyses
+
+        bool skipMergeEntry = false;
+        // Map callee(primal function) memory effects to the calling
+        // function's(autodiff op's) memory effects. This allows us to also
+        // reason about memory effects on the derivative argument potentially
+        // being passed in(if the primal argument has activity enzyme_dup)
+
+        // ForwardDiff read(primal) ?-> read(deriv);
+        // write(primal) ?-> write(deriv)
+        //
+        // ReverseDiff read(primal) ?-> read or write(deriv);
+        // write(primal) ?-> read or write(deriv);
+        //
+        // Unknown effects are retained in the caller, and will always alias to
+        // true with any other effect
+
+        SmallVector<MemoryEffects::EffectInstance, 4> callerEffects;
+
+        for (auto &eff : calleeEffects) {
+          if (isa<MemoryEffects::Read>(eff.getEffect()) ||
+              isa<MemoryEffects::Write>(eff.getEffect())) {
+            if (Value resource = eff.getValue()) {
+
+              auto argnum = 0;
+              bool found = false;
+              for (auto fnArg : key.function.getArguments()) {
+                if (fnArg == resource) {
+                  argnum = fnArg.getArgNumber();
+                  found = true;
+                  break;
+                }
+              }
+              auto e = eff.getEffect();
+              if (found) {
+
+                // Primal value as effect
+                Value primalVal = key.inputs[argnum];
+                if (auto primalOpResult = dyn_cast<OpResult>(primalVal))
+                  callerEffects.emplace_back(MemoryEffects::EffectInstance(
+                      eff.getEffect(), primalOpResult, eff.getResource()));
+                else if (auto primalBlockArg =
+                             dyn_cast<BlockArgument>(primalVal)) {
+                  callerEffects.emplace_back(MemoryEffects::EffectInstance(
+                      eff.getEffect(), primalBlockArg, eff.getResource()));
+                } else {
+                  llvm_unreachable("Value is neither an argument nor a result "
+                                   "of an op. This is not allowed by SSA");
+                }
 
-    // process map
-    for (auto mergeItr = toMerge.begin(); mergeItr != toMerge.end();
-         ++mergeItr) {
-      auto key = mergeItr->first;
-      SmallVector<enzyme::ForwardDiffOp> allOps = mergeItr->second;
-      auto width = allOps.size();
+                // Derivative effects(remain the same for fwddiff)
+                for (auto dop : allDiffs) {
```

Comment:
can you move this loop to precompute the mapping of getPrimalArgs() to be done once, outside of the call effects.

also frankly, we should jjust add a getPrimalArgs helper function to fwddiff/autodiff ops [and we can call that from here]